### PR TITLE
I1 library

### PR DESCRIPTION
**Pull request by amrmosit**
Refactored Library Class
Author Amr Mostafa
Reviewer Elliot
Madiator Haynes 

Hello Haynes, 
Here are my PR for Class Library after refactorying to conform with coding guidelines.

### CODE REVIEW

**Comment by haynescrossley** [comment]
Hey @amrmosit I can't seem to run Main successfully on your branch. Full error message: 

java.lang.RuntimeException: java.io.InvalidClassException: library.entities.Library; local class incompatible: stream classdesc serialVersionUID = -5667100925849200937, local class serialVersionUID = 6100019918649307782

If I delete the library.obj file on my end then it runs successfully. 

Not sure if this is an issue on my end or not.

@elliotmitchum let me know if you have the same?

**Comment by haynescrossley** [comment]
> Hey @amrmosit I can't seem to run Main successfully on your branch. You may have to delete your Library.obj file first? Full error message:
> 
> java.lang.RuntimeException: java.io.InvalidClassException: library.entities.Library; local class incompatible: stream classdesc serialVersionUID = -5667100925849200937, local class serialVersionUID = 6100019918649307782



> Refactored Library Class Author Amr Mostafa Reviewer Elliot Madiator Haynes
> 
> Hello Haynes, Here are my PR for Class Library after refactorying to conform with coding guidelines.

**Comment by haynescrossley** [comment]
Disregard the above. I can delete the .obj file on my end and it runs successfully. 

Ready for review @elliotmitchum.

File: src/library/entities/Library.java [hunk 1]
```diff
@@ -17,65 +17,67 @@
...
+                    self = (Library) libraryFile.readObject();
+                    Calendar.GeTiNsTaNcE().sEtDaTe(self.currentDate);
+                    libraryFile.close();
                 }
                 catch (Exception e) {
                     throw new RuntimeException(e);
                 }
             }
-            else SeLf = new Library();
+            else {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java:66]: Need to conform to bracket style. Should be:

```
if (...) {
  ...
} else {
  ...
}
```

File: src/library/entities/Library.java [hunk 2]
```diff
@@ -17,65 +17,67 @@
...
+                try (ObjectInputStream libraryFile = new ObjectInputStream(new FileInputStream(LIBRARY_FILE));) {
                 
-                    SeLf = (Library) LiBrArY_FiLe.readObject();
-                    Calendar.GeTiNsTaNcE().sEtDaTe(SeLf.CuRrEnT_DaTe);
-                    LiBrArY_FiLe.close();
+                    self = (Library) libraryFile.readObject();
+                    Calendar.GeTiNsTaNcE().sEtDaTe(self.currentDate);
+                    libraryFile.close();
                 }
                 catch (Exception e) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java:62]: Need to conform to bracket style. Should be:

```
try (...) {
  ...
} catch {
  ...
}
```

File: src/library/entities/Library.java [hunk 3]
```diff
@@ -17,65 +17,67 @@
...
-                LiBrArY_fIlE.close();    
+    public static synchronized void save() {
+        if (self != null) {
+            self.currentDate = Calendar.GeTiNsTaNcE().GeTdAtE();
+            try (ObjectOutputStream libraryFile = new ObjectOutputStream(new FileOutputStream(LIBRARY_FILE));) {
+                libraryFile.writeObject(self);
+                libraryFile.flush();
+                libraryFile.close();
             }
             catch (Exception e) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java:81]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038352

File: src/library/entities/Library.java [hunk 4]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
-        return NeXt_lOaN_Id++;
+    private long getNextLoanId() {
+        return nextLoanId++;
     }
 
     
-    public List<Patron> lIsT_PaTrOnS() {        
-        return new ArrayList<Patron>(PaTrOnS.values()); 
+    public List<Patron> listPatrons() {
+        return new ArrayList<Patron>(patrons.values());
```
Comment (elliotmitchum) [inline src/library/entities/Library.java:104]: Not sure if constructors count as methods in this instance?

> The argument list of any method calls MUST NOT contain any other method calls. E.g.

File: src/library/entities/Library.java [hunk 5]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
-        pAtRoN.TaKe_OuT_LoAn(loan);
-        iTeM.TaKeOuT();
-        LoAnS.put(loan.GeT_Id(), loan);
-        CuRrEnT_LoAnS.put(iTeM.GeTiD(), loan);
+    public Loan issueLoan(Item item, Patron patron) {
+        Date dueDate = Calendar.GeTiNsTaNcE().GeTdUeDaTe(LOAN_PERIOD);
+        Loan loan = new Loan(getNextLoanId(), item, patron, dueDate);
+        patron.TaKe_OuT_LoAn(loan);
+        item.TaKeOuT();
+        loans.put(loan.GeT_Id(), loan);
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: > The argument list of any method calls MUST NOT contain any other method calls. E.g.

Will need to do the folloing:

```
String id = loan.getId();
loans.put(loan.id, loan);
```

File: src/library/entities/Library.java [hunk 6]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+        currentLoans.put(item.GeTiD(), loan);
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038851

File: src/library/entities/Library.java [hunk 7]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
 
     
-    public double CaLcUlAtE_OvEr_DuE_FiNe(Loan LoAn) {
-        if (LoAn.Is_OvEr_DuE()) {
-            long DaYs_OvEr_DuE = Calendar.GeTiNsTaNcE().GeTDaYsDiFfErEnCe(LoAn.GeT_DuE_DaTe());
-            double fInE = DaYs_OvEr_DuE * FiNe_PeR_DaY;
-            return fInE;
+    public double calculateOverDueFine(Loan loan) {
+        if (loan.Is_OvEr_DuE()) {
+            long daysOverDue = Calendar.GeTiNsTaNcE().GeTDaYsDiFfErEnCe(loan.GeT_DuE_DaTe());
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038851

File: src/library/entities/Library.java [hunk 8]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+        patron.dIsChArGeLoAn(currentLoan);
+        item.TaKeBaCk(isDamaged);
+        if (isDamaged) {
+            patron.AdD_FiNe(DAMAGE_FEE);
+            damagedItems.put(item.GeTiD(), item);
         }
-        cUrReNt_LoAn.DiScHaRgE();
-        CuRrEnT_LoAnS.remove(itEM.GeTiD());
+        currentLoan.DiScHaRgE();
+        currentLoans.remove(item.GeTiD());
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038851

File: src/library/entities/Library.java [hunk 9]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
-        PAtrON.dIsChArGeLoAn(cUrReNt_LoAn);
-        itEM.TaKeBaCk(iS_dAmAgEd);
-        if (iS_dAmAgEd) {
-            PAtrON.AdD_FiNe(damageFee);
-            DaMaGeD_ItEmS.put(itEM.GeTiD(), itEM);
+        patron.dIsChArGeLoAn(currentLoan);
+        item.TaKeBaCk(isDamaged);
+        if (isDamaged) {
+            patron.AdD_FiNe(DAMAGE_FEE);
+            damagedItems.put(item.GeTiD(), item);
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038851

File: src/library/entities/Library.java [hunk 10]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+        }
     }
 
 
-    public void RePaIrITem(Item cUrReNt_ItEm) {
-        if (DaMaGeD_ItEmS.containsKey(cUrReNt_ItEm.GeTiD())) {
-            cUrReNt_ItEm.rEpAiR();
-            DaMaGeD_ItEmS.remove(cUrReNt_ItEm.GeTiD());
+    public void repairItem(Item currentItem) {
+        if (damagedItems.containsKey(currentItem.GeTiD())) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038851

File: src/library/entities/Library.java [hunk 11]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+            currentItem.rEpAiR();
+            damagedItems.remove(currentItem.GeTiD());
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038851

File: src/library/entities/Library.java [hunk 12]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
         }
-        else 
+        else {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java:242]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038283

File: src/library/fixitem/fIX_iTeM_cONTROL.java [hunk 13]
```diff
@@ -53,7 +53,7 @@ public void IteMInSpEcTeD(boolean mUsT_FiX) {
             throw new RuntimeException("FixItemControl: cannot call itemInspected except in INSPECTING state");
         
         if (mUsT_FiX) 
-            LiBrArY.RePaIrITem(CuRrEnT_ItEm);
+            LiBrArY.repairItem(CuRrEnT_ItEm);
```
Comment (elliotmitchum) [inline src/library/fixitem/fIX_iTeM_cONTROL.java:56]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934038283

File: src/library/entities/Library.java [hunk 14]
```diff
@@ -17,65 +17,67 @@
...
-        if (SeLf != null) {
-            SeLf.CuRrEnT_DaTe = Calendar.GeTiNsTaNcE().GeTdAtE();
-            try (ObjectOutputStream LiBrArY_fIlE = new ObjectOutputStream(new FileOutputStream(lIbRaRyFiLe));) {
-                LiBrArY_fIlE.writeObject(SeLf);
-                LiBrArY_fIlE.flush();
-                LiBrArY_fIlE.close();    
+    public static synchronized void save() {
+        if (self != null) {
+            self.currentDate = Calendar.GeTiNsTaNcE().GeTdAtE();
+            try (ObjectOutputStream libraryFile = new ObjectOutputStream(new FileOutputStream(LIBRARY_FILE));) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: I think the semicolon could be removed?

eg:

```
try (ObjectOutputStream libraryFile = new ObjectOutputStream(new FileOutputStream(LIBRARY_FILE))) {
```

Could also remove the extra line space after this statement.

File: src/library/entities/Library.java [hunk 15]
```diff
@@ -17,65 +17,67 @@
...
     
-    public static synchronized Library GeTiNsTaNcE() {        
-        if (SeLf == null) {
-            Path PATH = Paths.get(lIbRaRyFiLe);            
+    public static synchronized Library getInstance() {
+        if (self == null) {
+            Path PATH = Paths.get(LIBRARY_FILE);
             if (Files.exists(PATH)) {    
-                try (ObjectInputStream LiBrArY_FiLe = new ObjectInputStream(new FileInputStream(lIbRaRyFiLe));) {
+                try (ObjectInputStream libraryFile = new ObjectInputStream(new FileInputStream(LIBRARY_FILE));) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: See https://github.com/elliotmitchum/ITC205/pull/35/files#r934039699

File: src/library/entities/Library.java [hunk 16]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+    public List<Loan> listCurrentLoans() {
+        return new ArrayList<Loan>(currentLoans.values());
     }
 
 
-    public Patron aDd_PaTrOn(String firstName, String lastName, String email, long phoneNo) {        
-        Patron PaTrOn = new Patron(firstName, lastName, email, phoneNo, gEt_NeXt_PaTrOn_Id());
-        PaTrOnS.put(PaTrOn.GeT_ID(), PaTrOn);        
-        return PaTrOn;
+    public Patron addPatron(String firstName, String lastName, String email, long phoneNo) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: `phoneNo` needs to be meaningful and conform to class diagram:

To:

```
phoneNumber
```

File: src/library/entities/Library.java [hunk 17]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+        patrons.put(patron.GeT_ID(), patron);
+        return patron;
     }
 
     
-    public Item aDd_ItEm(String a, String t, String c, ItemType i) {        
-        Item ItEm = new Item(a, t, c, i, gEt_NeXt_ItEm_Id());
-        CaTaLoG.put(ItEm.GeTiD(), ItEm);        
-        return ItEm;
+    public Item addItem(String a, String t, String c, ItemType i) {
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: Parameter names need to be meaningful and conform to class diagram:

To:

```
String author, String title, String callNumber, ItemType itemType
```

File: src/library/entities/Library.java [hunk 18]
```diff
@@ -84,153 +86,155 @@ public static synchronized void SaVe() {
...
+        Item item = new Item(a, t, c, i, getNextItemId());
+        catalog.put(item.GeTiD(), item);
```
Comment (elliotmitchum) [inline src/library/entities/Library.java]: > The argument list of any method calls MUST NOT contain any other method calls. E.g.

To:

```
String id item.GeTiD();
catalog.put(id, item);
```

File: src/library/entities/Library.java [hunk 19]
```diff
@@ -17,65 +17,67 @@
 @SuppressWarnings("serial")
```
Comment (elliotmitchum) [inline src/library/entities/Library.java:17]: I've noticed that there are double line spaces for each method seperator, do we want to use a single or double line space for these? cc: @haynescrossley

**Review by elliotmitchum — CHANGES_REQUESTED** [review]
@amrmosit Have reviewed and left comments for changes and questions

File: src/library/entities/Library.java (hunk 1 above)
Comment (amrmosit) [reply src/library/entities/Library.java:66]: That is not the way it is in the CSG 8 -We agreed before on 
using the first option on Discord.
@elliotmitchum

File: src/library/entities/Library.java (hunk 2 above)
Comment (amrmosit) [reply src/library/entities/Library.java:62]: That is not included in CSG8 @elliotmitchum

File: src/library/entities/Library.java (hunk 3 above)
Comment (amrmosit) [reply src/library/entities/Library.java:81]: That is not included in CSG8. We agreed on using option number 1 @elliotmitchum

File: src/library/entities/Library.java (hunk 16 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: it's changed to phoneNumber to conform with CSG1. @haynescrossley @elliotmitchum 
PR 1 git commit -m"renaming variable phoneNumber in addPatron method to congorm with CSG1."
commit 029e0176240de619e8fcbc42858067aa8ff30f10

File: src/library/entities/Library.java (hunk 17 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: PR1-renamed parameter names for addItem method to conform with CSG1&2 @haynescrossley  @elliotmitchum 
commit[ q7914ada43ac3c323d0f07daab095f8eb4bf54e71](https://github.com/elliotmitchum/ITC205/pull/35/commits/7914ada43ac3c323d0f07daab095f8eb4bf54e71)

File: src/library/entities/Library.java (hunk 18 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: refining addItem method to conform with CSG10
commit 90e3f3eb870278e43d35182dbcfa567e41b22eee
@elliotmitchum @haynescrossley

File: src/library/entities/Library.java (hunk 5 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: refining issueLoan method to conform with CSG10
commit bba2e241634429347b510499af7b31ad4bdfc7ac
@haynescrossley @elliotmitchum

File: src/library/entities/Library.java (hunk 6 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: commit -m"refining issueLoan method to conform with CSG10"
commit 0b50da1306192740a58c230a3fd3cbf41f503499
@elliotmitchum  @haynescrossley

File: src/library/entities/Library.java (hunk 8 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: refining dischargeLoan method to conform with CSG10
commit 654e9ea9f12c74c0886969f545dccb2c318552f1
@elliotmitchum @haynescrossley

File: src/library/entities/Library.java (hunk 10 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: commit -m"refining repairItem method to conform with CSG10"
commit e01c4afba8a694b9fe5f9c2ffda1bc70cbc326e7 
@elliotmitchum  @haynescrossley

File: src/library/entities/Library.java (hunk 11 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: commit -m"refining repairItem method to conform with CSG10"
commit e01c4afba8a694b9fe5f9c2ffda1bc70cbc326e7 
@elliotmitchum  @haynescrossley

File: src/library/entities/Library.java (hunk 12 above)
Comment (amrmosit) [reply src/library/entities/Library.java:242]: same. CSG8. @elliotmitchum

File: src/library/entities/Library.java (hunk 7 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: commit -m"refining calculateOverDueFine method to conform with CSG10"
commit:[b8ed16c30c6395af72a6ff49619cbf062524335a](https://github.com/elliotmitchum/ITC205/pull/35/commits/158fe980e95fc99084abe5b714e8398fc17fb3d6)
@elliotmitchum @haynescrossley

File: src/library/entities/Library.java (hunk 9 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: refining dischargeLoan method to conform with CSG10
commit ab4f393b8daa0533b6f6b8c91f529ed19d778508
@elliotmitchum @haynescrossley

File: src/library/entities/Library.java (hunk 14 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: removing unnecessary semicolon and redundant close()
commit bc4aeab6316afbeca93f48394a7cdb3f349522bb
@elliotmitchum @haynescrossley

File: src/library/entities/Library.java (hunk 15 above)
Comment (amrmosit) [reply src/library/entities/Library.java]: commit -m"removing unnecessary semicolon"
commit d23e5080378e52102be407d5bbd569a41be7a06b
@elliotmitchum @haynescrossley

File: src/library/entities/Library.java (hunk 19 above)
Comment (amrmosit) [reply src/library/entities/Library.java:17]: I am not sure about this one.

File: src/library/entities/Library.java (hunk 4 above)
Comment (amrmosit) [reply src/library/entities/Library.java:104]: Not sure. @haynescrossley

File: src/library/entities/Library.java (hunk 19 above)
Comment (elliotmitchum) [reply src/library/entities/Library.java:17]: cc: @haynescrossley

File: src/library/entities/Library.java (hunk 4 above)
Comment (haynescrossley) [reply src/library/entities/Library.java:104]: Nothing specifically listed in CSG about constructors. Though I don't think there is any issue here. Its not a method, calling another method. Just returning the hash map values.

CSG example of what NOT to do:

someOtherVar = someOtherMethod(someMethod(x, y, z), p, q, r);

@amrmosit @elliotmitchum

**Review by elliotmitchum — APPROVED** [review]
Looks good to me

**Pull request merged**